File: ios/ThingzApp/Core/Network/APIClient.swift
import Foundation

enum APIClientError: Error {
    case invalidResponse
    case httpError(Int)
    case decodingError(Error)
    case tokenMissing
}

final class APIClient {
    static let shared = APIClient()
    private let session = URLSession.shared
    private let decoder = JSONDecoder()
    private init() {}

    func request<T: Decodable>(_ endpoint: APIEndpoint, as type: T.Type) async throws -> T {
        let data = try await send(endpoint)
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw APIClientError.decodingError(error)
        }
    }

    func requestVoid(_ endpoint: APIEndpoint) async throws {
        _ = try await send(endpoint)
    }

    private func send(_ endpoint: APIEndpoint) async throws -> Data {
        var request = URLRequest(url: endpoint.url)
        request.httpMethod = endpoint.method.rawValue
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        if endpoint.requiresAuth {
            guard let token = TokenManager.shared.accessToken else {
                throw APIClientError.tokenMissing
            }
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }

        for (key, value) in endpoint.customHeaders {
            request.setValue(value, forHTTPHeaderField: key)
        }

        if let body = endpoint.body {
            request.httpBody = try JSONEncoder().encode(body)
        }

        let (data, response) = try await session.data(for: request)

        guard let http = response as? HTTPURLResponse else {
            throw APIClientError.invalidResponse
        }

        if http.statusCode == 401, endpoint.requiresAuth {
            // 토큰 만료 시 재발급 시도
            try await refreshIfNeeded()
            return try await send(endpoint)
        }

        guard (200..<300).contains(http.statusCode) else {
            throw APIClientError.httpError(http.statusCode)
        }

        return data
    }

    private func refreshIfNeeded() async throws {
        guard let refreshToken = TokenManager.shared.refreshToken else {
            TokenManager.shared.clear()
            throw APIClientError.tokenMissing
        }
        let dto = try await request(.refresh(refreshToken: refreshToken), as: TokenDto.self)
        TokenManager.shared.save(dto)
    }
}

File: mobile/src/core/network/APIClient.ts
import TokenManager, { TokenDto } from '../auth/TokenManager';

const BASE_URL = 'http://192.168.0.160:8080';

type Method = 'GET' | 'POST' | 'DELETE';

async function request<T>(
  method: Method,
  path: string,
  options: { body?: object; headers?: Record<string, string>; auth?: boolean } = {}
): Promise<T> {
  const { body, headers = {}, auth = true } = options;

  if (auth) {
    const token = await TokenManager.getAccessToken();
    if (!token) throw new Error('TOKEN_MISSING');
    headers['Authorization'] = `Bearer ${token}`;
  }

  const res = await fetch(`${BASE_URL}${path}`, {
    method,
    headers: { 'Content-Type': 'application/json', ...headers },
    body: body ? JSON.stringify(body) : undefined,
  });

  if (res.status === 401 && auth) {
    await refreshTokens();
    return request(method, path, options);
  }

  if (!res.ok) throw new Error(`HTTP_${res.status}`);
  if (res.status === 204) return undefined as T;
  return res.json();
}

async function refreshTokens() {
  const refreshToken = await TokenManager.getRefreshToken();
  if (!refreshToken) { await TokenManager.clear(); throw new Error('TOKEN_MISSING'); }

  const dto: TokenDto = await request('POST', '/api/auth/refresh', {
    auth: false,
    headers: { 'X-Refresh-Token': refreshToken },
  });
  await TokenManager.save(dto);
}

export const api = {
  get:    <T>(path: string) => request<T>('GET', path),
  post:   <T>(path: string, body?: object, opts?: { auth?: boolean; headers?: Record<string, string> }) =>
    request<T>('POST', path, { body, ...opts }),
  put:    <T>(path: string, body?: object) => request<T>('PUT', path, { body }),
  patch:  <T>(path: string, body?: object) => request<T>('PATCH', path, { body }),
  delete: <T>(path: string, body?: object) => request<T>('DELETE', path, { body }),
};
